pennylane/vqe/vqe.py

```diff
@@ -120,6 +120,39 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def is_diagonal(self):
+        r""" Checks if a Hamiltonian is diagonal in the computational basis.
+
+        Returns:
+            bool: ``True`` if the Hamiltonian is diagonal in the computational basis, ``False`` otherwise.
+        """
+
+        non_diagonal_coeffs = []
+        non_diagonal_obs = []
+
+        # Loops through each term of the Hamiltonian
+        for i, term in enumerate(self._ops):
+
+            # Prunes identity from all tensor products
+            term = term.prune() if isinstance(term, qml.operation.Tensor) else term
+
+            # Combines like terms (provided they are non-diagonal)
+            if bool(np.count_nonzero(term.matrix - np.diag(np.diagonal(term.matrix)))):
```

Comment:
@Lucaman99 nice! And it will generalize to numeric observables as well.

However, a thought crosses my mind. This approach will work for `qml.Hermitian`, but calling `term.matrix` will be _very_ expensive for Hamiltonians on a large number of wires. Your other approach, where terms were simply matched against `qml.PauliZ` and `qml.Identity` scaled significantly better.

Would it be possible to only call `.matrix` on the `qml.Hermitian` terms directly, and continue to pattern match against `{qml.PauliZ, qml.Identity}` for the remaining observables?

Another approach would be to branch between this method and the old method if `qml.Hermitian` is in the term or not.

Reply:
Yes, I think this should be possible! One thing I was concerned about was the edge case in which the user defines a Hermitian that is equivalent to a Pauli gate. For instance:

``` python
H = qml.Hamiltonian([1, 1, -1], [qml.PauliZ(0), qml.Hermitian(np.array([[0, 1], [1, 0]), 1), qml.PauliX(1)])
print(H.is_diagonal())
```

should return `True`. However, I think I may have a solution! I believe there is a utility function that decomposes Hermitians into Pauli components, so we could use that, re-define the Hamiltonian, and then use the old logic to pattern match against diagonal gates. What do you think?

Reply:
I know the decomposition would be very expensive, but hopefully, cases in which the user is defining Hamiltonians in terms of massive, custom Hermitians is rare 😆

Reply:
I think this is a good plan! The thing I was more worried about was if the user has a Hamiltonian of the form

```python
qml.PauliZ(0) @ qml.PauliZ(1)  @ ...  @ qml.PauliZ(100) @ qml.PauliZ(101) @ qml.PauliZ(102)
```
etc. This is trivial to determine if it's diagonal using the operator matching approach, but will be a _huge_ computational cost to instead compute the dense matrix!